### ISSUE TITLE
Review requested automatically (and unexpectedly)

### ISSUE THREAD

**Issue by lassemaatta** [Reviewable/Reviewable#1184]
I'm not sure if this is due to a change in reviewable or github itself, but we recently noticed that _something_ has started to automatically request reviews from participants when a branch under review is altered.

What seems to happen is:
- Person A creates a PR
- Person B notices the PR in reviewable, adds some comments and publishes the review as normal
- Person A updates the branch locally and force pushes the changes, and we get a new revision in reviewable
- **We see "Person A requested review from Person B" event in github and an email is sent.**

Is this due to a change in reviewable? I couldn't find anything in the changelog.

**Comment by pkaminski** [comment]
Hey Lasse — no, we haven't made any recent changes in this area.  Reviewable will request reviewers in three situations:
1. When publishing from the client, if explicitly requested in a comment via `+reviewer:@username`.
2. When publishing from the client, if "[sync requested reviewers](https://docs.reviewable.io/reviews.html#requested-reviewers-synchronization)" is checked in the Publish dropdown.
3. When syncing the PR on the server, if the custom review completion condition sets [`syncRequestedReviewers`](https://docs.reviewable.io/repositories.html#syncrequestedreviewers).

That should be it, though.  Do any of these apply in your scenario?

**Comment by lassemaatta** [comment]
> Do any of these apply in your scenario?

I don't think so. I don't think I've ever seen or used the `+reviewer:@username` syntax in a comment. I found the "sync requested reviewers" checkbox in the Publish menu, but I don't believe I've ever used it. I tried looking at the repository settings (both in Github & reviewable) and found nothing related; we don't use any custom review completion conditions.

I noticed a similar event from yesterday:
- I wrote a PR
- My colleague left a comment but also approved the PR
- I wrote a reply to the comment (just plain text) and published it
- This was enough to trigger the "lassemaatta requested a review from $colleague" in github

I browsed through some earlier PR's at feels like this started somewhere around the middle of june.

**Comment by pkaminski** [comment]
Huh, we've got a bit of a mystery on our hands then.  To confirm, you see this happening when publishing comments but not on push and you don't have "sync requested reviewers" checked.  Right?  If so, could you check whether this also happens when:
1. You [send an individual message immediately](https://docs.reviewable.io/discussions.html#replying-to-a-discussion) instead of publishing.
2. You publish a review via GitHub.

Have you made any changes to your GitHub configuration recently, such as adding / changing your `CODEOWNERS` file perhaps?

**Comment by lassemaatta** [comment]
Right, so
- Unlike what I said in the original message, simply pushing new code does not trigger this
- Neither does sending an individual message instead of publishing
- Publishing a message _does_ trigger this and requests reviews from earlier commenters, even if the "sync requested reviews" is disabled

As far as I know we haven't changed our github configuration. Also, we are not using `CODEOWNERS`.

**Comment by pkaminski** [comment]
OK, so it clearly tracks back to the publishing flow.  Next time you're about to publish, could you gather a bit of data from your JS console?  Please run the following two commands and let me know the output of the second one:
```
review = truss.store.services.reviewContext.current.review;
{forced: review.forcedAutoRequestReviewers, settings: truss.store.users.current.settings.autoRequestReviewers, actual: review.state.autoRequestReviewers, effective: review.state.effectiveAutoRequestReviewers}
```

I don't see any way for publishing to affect requested reviewers unless `effective` is true, and if it is then hopefully the other values will let me know the root cause.  Thanks in advance!

**Comment by lassemaatta** [comment]
This is what I got (tested both Firefox & chrome) before publishing; unfortunately the results don't look very revealing:

`{forced: null, settings: undefined, actual: undefined, effective: false}`

**Comment by pkaminski** [comment]
Indeed.  Let's see if we can validate the inference that seeing the effect on publish means that the publishing flow is directly responsible.  If you know your way around the browser developer tools, could you check what requests to `github.com` are emitted when you publish?  We're specifically interested in anything with `requested_reviewers`.  Thanks.

**Comment by pkaminski** [comment]
Ping — did you get a chance to repro and check whether the browser is directly emitting a reviewer request?

**Comment by lassemaatta** [comment]
Yeah, I was just working on a suitable PR. When publishing some comments, I noticed
- A call to `https://api.github.com/repos/<group name>/<repo name>/compare/<commit 1>...<commit 2>?per_page=100` and

**Comment by pkaminski** [comment]
So the client is requesting the reviewer after all, even though the code path that would normally lead to this is closed per the earlier probe.  I have no idea what's going on but at least we're getting closer...  Now we need more data.

Immediately before publishing, please try this:
```
function render(directives) {return {reviewers: directives.reviewers, autoReviewers: directives.autoReviewers, overrides: _.map(directives.overrides, o => `${o.ref.toString()} => ${o.value}`)};}
review = truss.store.services.reviewContext.current.review;
render(await review.state.drafts._getDirectives(review.state.drafts.publishable));
```

Then Immediately after publishing, please try this:
```
render(_.findLast(Sentry.getCurrentHub()._stack[0].scope._breadcrumbs, {message: 'Analyzed directives'})?.data?.directives);
```

Thanks!

**Comment by pkaminski** [comment]
Hey, did you have any luck in getting a repro of this issue?  I'd still like to track it down.

(Note that we've update our Sentry integration so I edited some of the code in the message above.)

**Comment by lassemaatta** [comment]
Hi, unfortunately I've hadn't had the change to test this further. Next time I'm working on a suitable PR I'll try to gather the data you requested.

**Comment by pkaminski** [comment]
Ping — have you observed this misbehavior again?  We recently changed the way we post reviews to GitHub, and while I don't think it would affect this issue it's not impossible either.

**Comment by lassemaatta** [comment]
Hi, sorry for the lack of communication, it's been a busy few weeks. 

In any case, I can confirm the issue still occurs. I've been unable to gather the data you requested as either a) I forgot to do it (🤦‍♂️) or b) when I did remember it was one of those situations where it did not occur (e.g. the review was already requested previously). Still, I'll try to remember to get the data and hopefully get some clarity on this.

**Comment by lassemaatta** [comment]
I tested this again. I think something must have changed, because 

a) `review = truss.store.services.reviewContext.current.review;` threw an error about `truss` being undefined
b) the original issue does not seem to occur anymore

**Comment by pkaminski** [comment]
That's odd — we didn't change anything around the `truss` global and it still works for me on `reviewable.io`.  At this point it's more likely than not that we accidentally fixed this during our rework of the publishing process last month so I'll close the issue, but please reopen if you see it recur!
